### PR TITLE
CFE-656 - Removing --*oci-feature* flags

### PR DESCRIPTION
**Pull request by sherine-k**
# Description

As an end user I want use the "OCI FBC" feature without extra flags so that the user experience is simple and easy.

This PR is a Work In Progress. ToDos:
* Remove OriginalRef (which was required in 4.12 for OCI FBC catalogs) from the ImageSetConfig api
* Activate RenderDC for OCI FBC catalogs, in order to fix https://issues.redhat.com/browse/OCPBUGS-5085 
* Unit Tests

Fixes # [CFE-656](https://issues.redhat.com//browse/CFE-656)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

For the moment, a simple manual test has been done:
* Given that we copied the redhat-operator-index catalog locally (using skopeo for example) to `/tmp/oci-catalog2/`
* Use the following image set config 
```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
storageConfig:
  local:
    path: /tmp/storageBackend

mirror:
  platform:
    channels:
    - name: stable-4.12
      type: ocp
    graph: false 
  operators:
  #rhop
  - catalog: oci:///tmp/oci-catalog2/redhat-operator-index
    originalRef: registry.redhat.io/redhat/redhat-operator-index:v4.12am
    packages:
    - name: aws-load-balancer-operator
```
* Run 
```bashhttps://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax
make build
./bin/oc-mirror docker://localhost:5000/oci-tests --config imstcfg.yaml --dest-use-http --dest-skip-tls
```

# Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by lmzuccarelli — COMMENTED** [review]
Pulled the PR down locally.

Tested using this for imagesetconfig

```
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
storageConfig:
  local:
    path: /lmz-tmp/storageBackend
mirror:
  platform:
    channels:
    - name: stable-4.12
      type: ocp
    graph: false 
  operators:
  #rhop
  - catalog: oci:///lmz-tmp/oci-catalog
    originalRef: registry.redhat.io/redhat/redhat-operator-index:v4.12
    packages:
    - name: aws-load-balancer-operator
  additionalImages:
- name: registry.redhat.io/ubi8/ubi:latest
```

I created a fbc (oci) catalog using skopeo to local disk ( imagesetconfig  `/lmz-tmp/oci-catalog`)

The mirror did take a while to complete

The resulting `imageContentSourcePolicy` and `catalogueSource` look good

```
---
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: generic-0
spec:
  repositoryDigestMirrors:
  - mirrors:
    - localhost:5000/oci-tests/ubi8
    source: registry.redhat.io/ubi8
---
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: release-0
spec:
  repositoryDigestMirrors:
  - mirrors:
    - localhost:5000/oci-tests/openshift/release
    source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
  - mirrors:
    - localhost:5000/oci-tests/openshift/release-images
    source: quay.io/openshift-release-dev/ocp-release
```

**Comment by sherine-k** [comment]
/assign @jpower432 
Hi Jen
Would you mind giving us your opinion on this PR? We need your help to approve it. 
We're waiting also for feedback from @jchunkins 
Thanks!

File: pkg/cli/mirror/fbc_operators.go
Comment (jpower432) [inline pkg/cli/mirror/fbc_operators.go:368]: Nit: If the `digest` variable should always be set unless there was an error, it may be better to check for the digest variable first and error if that is not set. Might help catch issues in the case there was a uncaught error before calling this method.

File: pkg/cli/mirror/operator.go
Comment (jpower432) [inline pkg/cli/mirror/operator.go:157]: There are cases when `o.ToMirror` and o.UserNamespace` will be be set when it function is called. Is that handled earlier? In a `mirror-to-desk` workflow this would yield incomplete manifests or result in a failure.

**Review by jpower432 — COMMENTED** [review]
A few comments initially. Is this PR part of a planned series to achieve parity between remote and OCI based catalog input? I wonder if remove these feature flags would create more confusion between there is no differentiation between the two workflows. I know having to add the extra flag could seem cumbersome, but without them it could create confusion or unexpected behavior for the user. Thoughts?

**Review by jchunkins — CHANGES_REQUESTED** [review]
- Storage config is not being used and should be (its still defaulting to the current working directory)
   ```
   storageConfig:
     local:
       path: /tmp/localstorage <--- I expect this should be used for the workspace but its not doing that
   ```

- the `GetUniqueName` function in `pkg/api/v1alpha2/types_config.go` needs something to handle a OCI ref like the code below. This is necessary when the target name / tag are set, otherwise we get a parse error from `reference.Parse`, which causes the error message `error parsing source catalog` to be returned.
   ```go
   func (o Operator) GetUniqueName() (string, error) {
      if o.TargetName == "" && o.TargetTag == "" {
         return o.Catalog, nil
      }

      if strings.HasPrefix(o.Catalog, "oci:") { // <---- this check is needed
         return o.Catalog, nil
      }

      catalogRef, err := reference.Parse(o.Catalog)
      if err != nil {
         return "", fmt.Errorf("error parsing source catalog %s: %v", catalogRef, err)
      }
      if o.TargetName != "" {
         catalogRef.Name = o.TargetName
      }
      if o.TargetTag != "" {
         catalogRef.ID = ""
         catalogRef.Tag = o.TargetTag
      }

      return catalogRef.Exact(), nil
   }
   ```
- I attempted to generate a fat manifest using OCI layout and use it in the `ImageSetConfiguration` and could not get this to work. We would need to discuss this in person to illustrate the issue (it's too complicated to describe here).

- I am struggling with how to define my OCI path directory and the values to feed into the ImageSetConfiguration. 

   __Example command__:
   ```console
   oc-mirror -c ImageSetConfiguration.yml --oci-insecure-signature-policy --dest-skip-tls docker://localhost:5000
   ```
   __ImageSetConfiguration.yml file__:
   ```yaml
   kind: ImageSetConfiguration
   apiVersion: mirror.openshift.io/v1alpha2
   storageConfig:
   local:
      path: /tmp/localstorage
   mirror:
   operators:
   - catalog: oci:///tmp/ocitesting/amd64
      targetName: cp/test/common-services
      targetTag: v123
      packages:
      - name: ibm-auditlogging-operator-app
      channels:
      - name: v3
   ```

   __Results__:
   Catalog Source does not seem correct given the above input
   ```yaml
   apiVersion: operators.coreos.com/v1alpha1
   kind: CatalogSource
   metadata:
   name: ocitesting-amd64            # <- expecting this to use the target name (e.g. cp-test-common-services)
   namespace: openshift-marketplace
   spec:
   image: localhost:5000/ocitesting/cp/test/common-services:v123 # <- This is wrong see following discussion
   sourceType: grpc
   ```
   ICSP seems OK
   ```yaml
   ---
   apiVersion: operator.openshift.io/v1alpha1
   kind: ImageContentSourcePolicy
   metadata:
   labels:
      operators.openshift.org/catalog: "true"
   name: operator-0
   spec:
   repositoryDigestMirrors:
   - mirrors:
      - localhost:5000/opencloudio
      source: quay.io/opencloudio
   ```

   __Catalog Source__
   The catalog name and catalog source metadata name needs to reflect two different configuration paths (with `TargetName` and without `TargetName`), both of which end up using `ParseImageReference` to figure out paths. This function is used in a couple of places, but the path I am concerned with is the `prepareDestCatalogRef` function. So `ParseImageReference` function assumes the first element in the path is the "registry", the last element in the path is the "repository" (i.e. the image name) and everything in between is the "organization" (i.e. a namespace made up of zero or more paths). What if I want to put my catalog in some arbitrary location (i.e. not at the root of the file system)? If I do something like the following example where I've used the name "registry", and "repository" as the parts of the path to illustrate the beginning end of the actual reference. In other words, I have a directory where I want to store my OCI artifacts which is `/some/arbitrary/path` and in that directory I have the actual OCI path reference `registry/some/value/for/org/repository`:

   ```go
   registry, organization, repository, tag, digest := ParseImageReference("oci:///some/arbitrary/path/registry/some/value/for/org/repository")
   // results:
   //   registry = "some"
   //   organization = "arbitrary/path/registry/some/value/for/org"
   //   repository = "repository"
   //   tag = ""
   //   digest = ""
   ```
   As you can see, the first part of the oci reference resolves to "some" (rather than "registry") and the repository resolves to "repository". There's no way to put the OCI artifacts in any location other than the root if I want to get the "registry" value to resolve correctly (e.g. I would need to use `oci:///registry/some/value/for/org/repository`). I don't see a way to handle arbitrary file paths for an OCI image and to be able to parse this in any sensible way without adding yet another field to indicate where the OCI images will live (so you could strip off that portion from the path). Adding new fields is something I'd like to avoid. So here's my proposal:


   1. when the catalog DOES NOT have the `operators[].targetName` defined
      - if the `operators[].catalog` field refers to an OCI image, AND `operators[].targetName` is not defined, this is an error condition (i.e. `operators[].targetName` is required in this case)
   1. when the catalog DOES have the `operators[].targetName` defined
      - the last part of the `operators[].targetName` is the "repository"
      - everything before "repository" is the "org" (if present)
      - registry SHOULD NOT be included in the `operators[].targetName`

   So this allows us to use an OCI reference for what it is... a path to the file system where the actual OCI content lives. We don't try to parse that value in any way since it does not make sense to do so. The `operators[].targetName` becomes the way you configure what the final docker reference is going to look like when you push it to its final destination.

**Comment by sherine-k** [comment]
> A few comments initially. Is this PR part of a planned series to achieve parity between remote and OCI based catalog input? I wonder if removing these feature flags would create more issues because there is no differentiation between the two workflows. I know having to add the extra flags could seem cumbersome, but without them it could create confusion or unexpected behavior for the user. Thoughts?

@jpower432 You might have a point.
We also were concerned that adding OCI FBC catalogs in the ImageSetConfig would result in mixing (according to the command line used) a DiskToMirror workflow for the OCI catalog(s) and  MirrorToMirror workflow for everything else.
 I propose that we talk about it in tomorrow's meeting (invitation sent and meet https://meet.google.com/udc-mqrv-cpb) if you have time

File: pkg/api/v1alpha2/types_config.go
Comment (sherine-k) [reply pkg/api/v1alpha2/types_config.go:119]: Some context: 
`OriginalRef` was added in 4.12 as part of the TechPreview on OCI FBC catalogs.
We didn't increment the api version at the time. 
IBM requested its removal for 4.13.
Waiting for @dinhxuanvu and @soltysh's feedback

**Comment by sherine-k** [comment]
> Storage config is not being used and should be (its still defaulting to the current working directory)
I stand to be corrected, but `oc-mirror-workspace` has always been a folder created in the current directory. 
@jpower432 can you confirm?
 The Storage config is a place to store the metadata for mirroring (history of previous mirrors). In the tests I did, `/tmp/localstorage` was a `publish` folder containing `.metadata.json`, same as in 4.10 version of oc-mirror.

**Comment by sherine-k** [comment]
> Results:
Catalog Source does not seem correct given the above input

You're right that we are not using the targetName for the `metadata.name`.
One thing we didn't expect is the use of `/` in the `targetName`.
What we can do, is report an error if we see `/` in `targetName`. 
So a targetName that might be accepted is `common-services` for example.
What we should expect then is : 
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
name: common-services            
namespace: openshift-marketplace
spec:
image: localhost:5000/ocitesting/common-services:v123 
sourceType: grpc
```

For the use case that you are describing, let's discuss this live.
We already know from the slack forum that other users are asking for custom namespaces or paths for the catalogs mirrored. And for now, this is a big change, with a lot of impact on all existing workflows,  if we start having targetName used differently for oci and non-oci.
 Can we discuss it live and find a better solution?
@jpower432 , @afflom, @heheffne we're going to need your help here.

File: pkg/image/mapping.go
Comment (sherine-k) [reply pkg/image/mapping.go:23]: It's up for discussion: 
`OtherFormat` is to say that no analysis into the image is done to see if it's OCI, DockerV2S2 or DockerV2S1 or anything else. 
This is why it's used almost always, unless we explicitely find an OCI FBC...
Later, we might be interested (or not, I personally don't know) in pinpointing the format more specifically.
What do @jpower432 @lmzuccarelli @dinhxuanvu think?

File: pkg/cli/mirror/operator.go
Comment (sherine-k) [reply pkg/cli/mirror/operator.go:157]: I added a condition here. I'm not sure if there's a better error message to display

File: pkg/cli/mirror/fbc_operators.go
Comment (sherine-k) [reply pkg/cli/mirror/fbc_operators.go:368]: Makes sense.
Thanks @jpower432

File: pkg/image/mapping.go
Comment (sherine-k) [reply pkg/image/mapping.go:17]: Added :+1:

**Comment by sherine-k** [comment]
> the GetUniqueName function in pkg/api/v1alpha2/types_config.go needs something to handle a OCI ref like the code below. 

Fixed. PTAL

File: pkg/image/mapping.go
Comment (jpower432) [reply pkg/image/mapping.go]: Would using something like an enum type be better or are these string value being used somewhere? You could just allow the default to be set at something like "unset" or "invalid". The only concern I have for setting the default to `DockerV2" is not being able to detect if the value is set or unset.

**Comment by jpower432** [comment]
> > Storage config is not being used and should be (its still defaulting to the current working directory)
> 
> I stand to be corrected, but `oc-mirror-workspace` has always been a folder created in the current directory. @jpower432 can you confirm? The Storage config is a place to store the metadata for mirroring (history of previous mirrors). In the tests I did, `/tmp/localstorage` was a `publish` folder containing `.metadata.json`, same as in 4.10 version of oc-mirror.

@sherine-k This is correct. `storageConfig` is for metadata storage and not for the workspace. The default location of the workspace is `oc-mirror-workspace` in the local directory or the directory specified when using a file destination. There is a hidden flag that can change this https://github.com/openshift/oc-mirror/blob/74d455f5f6f3ab1b41127172d98005b46db2d7ca/pkg/cli/options.go#L27.

**Comment by jpower432** [comment]
> > Results:
> > Catalog Source does not seem correct given the above input
> 
> You're right that we are not using the targetName for the `metadata.name`. Fixed with
> 
> https://github.com/openshift/oc-mirror/blob/71ebe131ff3726a8ca95987244bf48fc9c3c91d3/pkg/cli/mirror/fbc_operators.go#L358
> 
> 
> One thing we didn't expect is the use of `/` in the `targetName`.
> What we can do, is report an error if we see `/` in `targetName`.
> So a targetName that might be accepted is `common-services` for example.
> What we should expect then is :
> ```
> apiVersion: operators.coreos.com/v1alpha1
> kind: CatalogSource
> metadata:
> name: common-services            
> namespace: openshift-marketplace
> spec:
> image: localhost:5000/ocitesting/common-services:v123 
> sourceType: grpc
> ```
> 
> For the use case that you are describing, let's discuss this live. We already know from the slack forum that other users are asking for custom namespaces or paths for the catalogs mirrored. And for now, this is a big change, with a lot of impact on all existing workflows, if we start having targetName used differently for oci and non-oci. Can we discuss it live and find a better solution? @jpower432 , @afflom, @heheffne we're going to need your help here.

@sherine-k there is some code that does `ImagesetConfiguration` validatation where you could add some use cases for these inputs https://github.com/openshift/oc-mirror/blob/74d455f5f6f3ab1b41127172d98005b46db2d7ca/pkg/config/validate.go#L16.

**Comment by sherine-k** [comment]
> I am struggling with how to define my OCI path directory and the values to feed into the ImageSetConfiguration.

As discussed in yesterday's meeting, here's what we can use:
Example command:

oc-mirror -c ImageSetConfiguration.yml --oci-insecure-signature-policy --dest-skip-tls docker://localhost:5000/cp/test

ImageSetConfiguration.yml file:
```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
storageConfig:
local:
   path: /tmp/localstorage
mirror:
operators:
- catalog: oci:///tmp/ocitesting/amd64
   targetName: common-services
   targetTag: v123
   packages:
   - name: ibm-auditlogging-operator-app
   channels:
   - name: v3
```
Results:
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
name: common-services           
namespace: openshift-marketplace
spec:
image: localhost:5000/cp/test/ocitesting/common-services:v123 
sourceType: grpc

---
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  labels:
    operators.openshift.org/catalog: "true"
  name: operator-0
spec:
  repositoryDigestMirrors:
  - mirrors:
    - localhost:5000/cp
   source: quay.io/opencloudio
```

The ICSP isn't perfect: the correct behavior would be :
```
---
apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  labels:
    operators.openshift.org/catalog: "true"
  name: operator-0
spec:
  repositoryDigestMirrors:
  - mirrors:
    - localhost:5000/cp/test/opencloudio
   source: quay.io/opencloudio
```
We'll tend to this after talking with maintainers of [library-go](https://github.com/openshift/oc-mirror/blob/74d455f5f6f3ab1b41127172d98005b46db2d7ca/vendor/github.com/openshift/library-go/pkg/image/reference/reference.go#L13), where the parsing of file based, and non file based image references happens.

**Comment by sherine-k** [comment]
Closing this PR based on the discussion from Jan 31st. 
See https://docs.google.com/document/d/1WAlggMit6ui0Y40BeSXwcUZg-LhretUZtv__G3dp8WY/edit# 
The decision was made to keep the flags, in order to shield this oci fbc feature, which is very specific and should not be exposed as the normal usage of oc-mirror.

**Pull request closed without merging**
